### ISSUE TITLE
Empty commits since 0.25.0

### ISSUE THREAD

**Issue by mweinelt** [ytti/oxidized#1666]
For a single host (out of ~100 hosts) in a single repository I'm experiencing empty commits on every update run since I've upgraded oxidized to 0.25.0. 

I'm not sure what other information is helpful so go ahead an ask.

**Comment by wk** [comment]
At minimum, it would be useful to include the model of the affected device and a debug log of device interaction acquired by enabling [debugging](https://github.com/ytti/oxidized/blob/master/docs/Configuration.md#debugging) with both a healthy device and the problem device, if the same model is used on more than the affected device itself.

Including a copy of the oxidized log from a run in which both affected and non-affected devices are included could also assist in pinpointing the issue.

**Comment by LarsKollstedt** [comment]
Am Freitag, 11. Januar 2019, 20:24:43 CET schrieb Wild Kat:
> At minimum, it would be useful to include the model of the affected device
> and a debug log of device interaction acquired by enabling
> [debugging](https://github.com/ytti/oxidized/blob/master/docs/Configuration
> .md#debugging) with both a healthy device and the problem device, if the
> same model is used on more than the affected device itself.
> 
> Including a copy of the oxidized log from a run in which both affected and
> non-affected devices are included could also assist in pinpointing the
> issue.

Hi,

the affected device edge1.e0.dolivo11.da.man-da.net is a Juniper ACX2100-AC 
like e.g. edge1.tiz.da.man-da.net.


Find the output of 

egrep -A 30 -B 30 'edge[0-9]' /var/log/oxidized.log | sed -E "s/node key 
'password' (to|with) value '[^']+'/node key 'password' \1 value '<password>'/
g" > ~/oxidized-filtered.log

attached. I also replaced the username afterwards. The edge\d\..* devices are 
all of the types Juniper ACX2100-AC and ACX1100-AC and running a JUNOS 16.2R2-
S3. For nearly half a year without oxidized issues. edge1.e0.doliv11.da was 
the first device, upgraded from 12.3R10.2 on 2018-06-01. But that's the only 
special thing I found.
There was no configuration change on the affected device surrounding the start 
of the empty commits.

Debug is set to true, but I can't see a difference between this devices. In 
the debug output. Or did I overlook something?


This was one of the first things I had a look at. 

I also did some troubleshooting in the GIT direction:
  - having a look at meta data (e.g. file permissions) in a checkout clone 
    from the repository on our internal gogs server an from the repository in 
    oxidized's workplace. Nothing special.
  - Doing a real change on the device, to force change in the dumped config.
  - delete the device's file from the repository (git obviously notices it's 
    still the same file)
  - deleting and bare cloning the repository to oxidized workplace again

We also filtered out the device in the SQL query. No other device showed this. 
Removing the filter lets this occur again for the same device. The Oxidized-
Backup has been disabled for approx 2 Weeks for this device, now. And this 
happens again today when I removed the filter again, after your question.


The hottest trail we found til, now. Are the

commit d95084059f20c04e109b0843b2ce8cd3c6abf770 created on 3 Nov 2018

and

commit 415e0318a49290ff592132de2ad1c916f8c8e337 created on Nov 12, 2018

in oxidized but and also my colleague Martin are not deep enough into Ruby and 
Rugged to really deep understand how the commits are created and avoided 
there.


And we of course don't understand why this works for all but not for this 
device. The device name is a long one, but not the longest we use.

Race conditions with such persistance are rare (e.g. colliding internal IDs 
derivated from something?). But I haven't seen any place where is something 
like this done in oxidized, yet? The error first occurred immediately after 
the upgrade to 0.25.0 on Dec 17 11:15. And rugged was last updated on Dec 10 
14:07.

And that was the end of my ideas.

Kind regards,
	Lars


-- 
Lars Kollstedt

Telefon: +49 6151 16-71027
E-Mail:  lk@man-da.de

man-da.de GmbH
Dolivostraße 11
64293 Darmstadt

Sitz der man-da.de GmbH: Darmstadt
Amtsgericht Darmstadt, HRB 9484
Geschäftsführer: Andreas Ebert

**Comment by LarsKollstedt** [comment]
Since attachement didn't suvive, here it is.
[oxidized-filtered.log](https://github.com/ytti/oxidized/files/2751483/oxidized-filtered.log)

**Comment by ytti** [comment]
We had wrong file permissions in 0.25.0 so 0.25.1 was released shortly after, but it's not immediately obvious to me how it would result in empty config.

What if you manually run 'ssh router "show version"', does that work? Because that is what we do, we run ssh exec channel for each of the command. 

If that does work and produces useful output, would it be possible to provide sanitised version of non-working device online, so I can debug it?

**Comment by LarsKollstedt** [comment]
Hi,

on Saturday, 12. Januar 2019, 01:18:00 CET you wrote:
> We had wrong file permissions in 0.25.0 so 0.25.1 was released shortly
> after, but it's not immediately obvious to me how it would result in empty
> config.
If you mean #1661 with this. I read this as an permission issue on the gem 
folder. And I meant a permission or meta data issue within the git checkout. I 
don't think that has something to do with each other.

We updated to 0.25.0 and a day later to 0.25.1, but the empty commits started 
with 0.25.0. And I at least don't know about a specific event to 
edge1.e0.dolivo11.da during the short day we really were on 0.25.0.


> What if you manually run 'ssh router "show version"', does that work?
> Because that is what we do, we run ssh exec channel for each of the
> command.
Yes it does. For the affected device edge1.e0.dolivo11.da.man-da.net as well 
as for other ACX2100-ACs. The output looks usefull and like the same.

When I do real changes to edge1.e0.dolivo11.da they are also backuped 
correctly. But afterwards it goes on with empty commits. There was no change 
neither of the dumped configuration nor of the configuration on the device 
surrounding the upgrade.


> If that does work and produces useful output, would it be possible to
> provide sanitised version of non-working device online, so I can debug it?
It probably would be possible for me to deploy a oxidized-Server that is 
backuping only one ACX2100 where it works and the affected one from a subnet 
with another security context than the subnet of the current oxidized server.

If the the error stays on edge1.e0.dolivo11.da.man-da.net there, it would be 
possible to give you access to that test server using a ssh-public-key.


But I won't be able to do this before Monday.


The user which is used by the oxidized process for that is only allowed to do 
a few commands on the backuped devices by tacacs+, but I'll use a clone of it 
with a different password and the same settings for that purpose.

If that wouldn't reproduce the issue, I'll have to find the relevant 
difference. But we'll probably have a new trail to follow.
   
Kind regards,
	Lars

-- 
Lars Kollstedt

Telefon: +49 6151 16-71027
E-Mail:  lk@man-da.de

man-da.de GmbH
Dolivostraße 11
64293 Darmstadt

Sitz der man-da.de GmbH: Darmstadt
Amtsgericht Darmstadt, HRB 9484
Geschäftsführer: Andreas Ebert

**Comment by LarsKollstedt** [comment]
Hi,

when I consolidated the config of the two devices for the Setup described above a bit (by doing a diff between them), i fell over an non 7bit ASCII character (an "ü") in an interface description. Normally we rewrite this one to "ue", but on edge1.e0.dolivo11.da it was part of the device config. And it was obviously the only non 7bit ASCII character.
This "ü" was first introduced on November 22th 2017, and was backuped correctly all times. But since we installed oxidized 0.25.0 on December 17th 2018 11:30 CET, the configuration obviously always was threaded as differing.
When I rewrite the "ü" to "ue" in the device configuration the empty commits are stopping and the "ü" is also rewritten to "ue" in the backuped config. When I reintroduce it, it's written correctly to the backup-ed configuration again, but the empty commits are starting again.

I've rewritten it now. And the empty commits commits are gone for now. But this was formerly working correctly. ;-)

Kind regards,
       Lars

**Comment by LarsKollstedt** [comment]
Hi,

on Tuesday, January 15th 2019, 09:08:27 CET you:
> I would love to take a peek on affected device to address the issue. At some
> point we introduced ability to tread strings as 8bit data or force UFC8
> conversion, it maybe that particular code is broken and returns empty upon
> conversion.

Can you please send me a ssh-publickey for that?


My first thought on the results are, possibly the comparison happens before 
the conversion, now. But I don't really know the code at that point, so I 
can't really say what you are comparing in this case, now. Comparing an empty 
or an cut of config will result in the same.

All I can say at this point is, the conversion is obviously done for the saved 
data.

Kind regards,
	Lars

-- 
Lars Kollstedt

Telefon: +49 6151 16-71027
E-Mail:  lk@man-da.de

man-da.de GmbH
Dolivostraße 11
64293 Darmstadt

Sitz der man-da.de GmbH: Darmstadt
Amtsgericht Darmstadt, HRB 9484
Geschäftsführer: Andreas Ebert

**Comment by LarsKollstedt** [comment]
Am Montag, 28. Januar 2019, 13:26:21 CET schrieb ytti:
> ssh-rsa
[...]

Hi,

the Key is now placed on oxidized-test.man-da.de.

Identification of this machine is:

oxidized-test.man-da.de,82.195.72.5,2001:41b8:200::1:5 ecdsa-sha2-nistp256 
AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBNFnN0u4l5uOvfu0tN1pSmLB/
mQb6JpK4ypmsocLSV7xmFCxgLm62Y5vmUNC0qVlSBSV7CpelDQNaORz8a5bYrI=
oxidized-test.man-da.de,82.195.72.5,2001:41b8:200::1:5 ssh-ed25519 
AAAAC3NzaC1lZDI1NTE5AAAAIAnurI5kDlkfPNm9SpF/FrWNfU0vXPnSOpQUpICYiEef
oxidized-test.man-da.de,82.195.72.5,2001:41b8:200::1:5 ssh-rsa 
AAAAB3NzaC1yc2EAAAADAQABAAABAQCdw8TWBsG1qd7Z7OX0kmDOlwycJ9dMnpCGJ8hybyP3fStQYOQm1/
iqlC/oZoAsIGdJODIHug6fgZMOq
+KUHYdJhLaZXZz40tda2T3eWwkWtuY2WoZrbdU8DxV4v5RmCXem6hWKD6bZACftIuLZDPV7FuPOzaB
+mG8lCS6NIgwoMrfNUQP5qoLiAII/rx2ej1arOzz+0RVTrBIH1J8cGMUZACTyFbDBDixLRqXAaJQ/
WeaR8wFTCP22XDzC5hHG3bZ3ehHzE/SA1OKL6OBGCbPT7iVUqyU2oyOMjLAx0LXAXxH/dKDFT/
yy8JqP7RL8JiUWfCK36ZCWXNlFsGmlCpaB


I gave you sudo for some oxidized related stuff at first:
ytti ALL=(ALL) NOPASSWD: /bin/systemctl start oxidized
ytti ALL=(ALL) NOPASSWD: /bin/systemctl stop oxidized
ytti ALL=(ALL) NOPASSWD: /bin/systemctl restart oxidized
ytti ALL=(ALL) NOPASSWD: /bin/systemctl status oxidized
ytti ALL=(ALL) NOPASSWD: /bin/journalctl -u oxidized

And you have full write access to all files and directories under /var/lib/
gems where oxidized is installed. Feel free to put your debug outputs in.


The log is written to /var/log/oxidized/oxidized.log, and the username the 
test host uses to backup can be found in /etc/oxidized/config. It's restricted 
to a limited set of show commands.

The backups are written to /home/oxidized/darmstadt.git, I used /home/
oxidized/darmstadt-checkout to have a look into this.

I reintroduced the ü on edge1.e0.dolivo11.da, now. And hope I've forgotten 
nothing important. ;-)

Kind regards,
	Lars

-- 
Lars Kollstedt

Telefon: +49 6151 16-71027
E-Mail:  lk@man-da.de

man-da.de GmbH
Dolivostraße 11
64293 Darmstadt

Sitz der Gesellschaft: Darmstadt
Amtsgericht Darmstadt, HRB 9484
Geschäftsführer: Andreas Ebert
